### PR TITLE
eos 0.12.1 with hunterizations

### PR DESCRIPTION
**Pull request by headupinclouds**
Re-apply hunterizations to eos 0.12.1 CMakeLists.txt
  * add HunterGate
  * use `project(.... VERSION ...)` instead of separate variable set
  * replace submodule use with simple `hunter_add_package + find_package`
  * add boilerplate package config installation
  * use `find_dependency` calls in Config.cmake.in template
  * hunterize examples

Hunterized build works.

Caveat: `EOS_BUILD_CERES_EXAMPLE=ON` requires hunter glm v0.9.8 (per eos issue report).  PR in progress.

python (pybind11) and matlab (mexplus) submodules/bindings ignored.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,122 +1,117 @@
...
 
-# Find dependencies:
 hunter_add_package(OpenCV)
 find_package(OpenCV REQUIRED core)
-message(STATUS "OpenCV include dir found at ${OpenCV_INCLUDE_DIRS}")
-message(STATUS "OpenCV lib dir found at ${OpenCV_LIB_DIR}")
 
 hunter_add_package(Boost COMPONENTS system)
-find_package(Boost CONFIG REQUIRED system)
+find_package(Boost 1.50.0 COMPONENTS system REQUIRED)
```
Comment (ruslo) [inline CMakeLists.txt]: Should be `find_package(Boost CONFIG REQUIRED system)`

File: cmake/Config.cmake.in [hunk 2]
```diff
@@ -1,10 +1,22 @@
 @PACKAGE_INIT@
 
+<<<<<<< HEAD
```
Comment (ruslo) [inline cmake/Config.cmake.in]: Looks like unresolved conflict

File: cmake/HunterGate.cmake [hunk 3]
```diff
@@ -88,8 +88,7 @@ endfunction()
 
 function(hunter_gate_fatal_error)
   cmake_parse_arguments(hunter "" "WIKI" "" "${ARGV}")
-  string(COMPARE EQUAL "${hunter_WIKI}" "" have_no_wiki)
-  if(have_no_wiki)
+  if(NOT hunter_WIKI)
```
Comment (ruslo) [inline cmake/HunterGate.cmake]: HunterGate is not the latest

File: examples/CMakeLists.txt [hunk 4]
```diff
@@ -1,38 +1,37 @@
...
-endif()
-message(STATUS "OpenCV include dir found at ${OpenCV_INCLUDE_DIRS}")
-message(STATUS "OpenCV lib dir found at ${OpenCV_LIB_DIR}")
-# This allows us to compile in RelWithDebInfo. It'll use the Release-version of OpenCV:
-set_target_properties(${OpenCV_LIBS} PROPERTIES MAP_IMPORTED_CONFIG_RELWITHDEBINFO RELEASE)
+find_package(OpenCV REQUIRED core highgui)
 
 hunter_add_package(Boost COMPONENTS system filesystem program_options)
-find_package(Boost CONFIG REQUIRED system filesystem program_options)
+find_package(Boost 1.50.0 COMPONENTS system filesystem program_options REQUIRED)
```
Comment (ruslo) [inline examples/CMakeLists.txt]: Should be `find_package(Boost CONFIG REQUIRED system filesystem program_options)`

File: examples/CMakeLists.txt [hunk 5]
```diff
@@ -1,38 +1,37 @@
...
+message(STATUS "Boost found at ${Boost_INCLUDE_DIRS}")
 
-if(MSVC)
-	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /bigobj")
-endif()
+# Simple model fitting (orthographic camera & shape to landmarks) example:
+add_executable(fit-model-simple fit-model-simple.cpp)
+target_link_libraries(fit-model-simple eos ${OpenCV_LIBS} ${Boost_LIBRARIES})
```
Comment (ruslo) [inline examples/CMakeLists.txt]: `s,Boost_LIBRARIES,Boost::system Boost::filesystem Boost::program_options,`

File: examples/CMakeLists.txt [hunk 6]
```diff
@@ -1,38 +1,37 @@
...
+target_link_libraries(fit-model-simple "$<$<CXX_COMPILER_ID:GNU>:-pthread>$<$<CXX_COMPILER_ID:Clang>:-pthreads>")
```
Comment (ruslo) [inline examples/CMakeLists.txt]: We can use `Threads::Threads` here, see:
* https://cmake.org/cmake/help/latest/module/FindThreads.html

File: examples/CMakeLists.txt [hunk 7]
```diff
@@ -1,38 +1,37 @@
...
+
+if(EOS_BUILD_CERES_EXAMPLE)
+  # Find Ceres, for the fit-model-ceres app:
+  hunter_add_package(ceres-solver)
+  find_package(Ceres CONFIG REQUIRED)
+  message(STATUS "Ceres locations: Headers: ${CERES_INCLUDE_DIRS} Library: ${CERES_LIBRARIES}")
+
+  # Single and multi-image non-linear model fitting with Ceres example:
+  add_executable(fit-model-ceres fit-model-ceres.cpp)
+  target_link_libraries(fit-model-ceres eos ${CERES_LIBRARIES} ${OpenCV_LIBS} ${Boost_LIBRARIES})
```
Comment (ruslo) [inline examples/CMakeLists.txt]: It should be just `ceres` instead of `CERES_LIBRARIES`:
* https://github.com/ruslo/hunter/wiki/pkg.ceres-solver

**Review by ruslo — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (headupinclouds) [reply CMakeLists.txt]: 👍

File: cmake/Config.cmake.in (hunk 2 above)
Comment (headupinclouds) [reply cmake/Config.cmake.in]: 👍

File: cmake/HunterGate.cmake (hunk 3 above)
Comment (headupinclouds) [reply cmake/HunterGate.cmake]: Updated.

File: examples/CMakeLists.txt (hunk 4 above)
Comment (headupinclouds) [reply examples/CMakeLists.txt]: 👍

File: examples/CMakeLists.txt (hunk 5 above)
Comment (headupinclouds) [reply examples/CMakeLists.txt]: Thanks.  Added `set(boost_libs Boost::system Boost::filesystem Boost::program_options)` for reuse.

File: examples/CMakeLists.txt (hunk 6 above)
Comment (headupinclouds) [reply examples/CMakeLists.txt]: Thanks. 

```
# https://stackoverflow.com/a/39547577
set(CMAKE_THREAD_PREFER_PTHREAD TRUE)
set(THREADS_PREFER_PTHREAD_FLAG TRUE)
find_package(Threads REQUIRED)

# Simple model fitting (orthographic camera & shape to landmarks) example:
add_executable(fit-model-simple fit-model-simple.cpp)
target_link_libraries(fit-model-simple eos ${OpenCV_LIBS} ${boost_libs})
target_link_libraries(fit-model-simple Threads::Threads)
install(TARGETS fit-model-simple DESTINATION bin)
```

File: examples/CMakeLists.txt (hunk 7 above)
Comment (headupinclouds) [reply examples/CMakeLists.txt]: Fixed.

**Comment by headupinclouds** [comment]
Thanks for the review.  Much cleaner.

The `glm` update fixed the `ceres` feature.

```
CONFIG=Release; polly.py --install --toolchain xcode --config ${CONFIG} --verbose --fwd HUNTER_CONFIGURATION_TYPES=${CONFIG} EOS_BUILD_CERES_EXAMPLE=ON --reconfig --test
```
```
...
Generate: 0:02:59.657419s
Build: 0:00:28.284248s
Test: 0:00:00.052438s
Total: 0:03:27.994629s
SUCCESS
```

```
$ cd _install/xcode/bin/
$ ./fit-model-simple 
Finished fitting and wrote result mesh and isomap to files with basename "out".
$ ./fit-model
Finished fitting and wrote result mesh and isomap to files with basename "out".
```

Magic!

out

File: matlab/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+  find_package(OpenCV 3 REQUIRED core imgproc imgcodecs)
+endif()
+set_target_properties(${OpenCV_LIBS} PROPERTIES MAP_IMPORTED_CONFIG_RELWITHDEBINFO RELEASE)
+
+if(MSVC)
+  # The standard find_package for boost on Win finds the dynamic libs, so for dynamic linking to boost we need to #define:
+  add_definitions(-DBOOST_ALL_NO_LIB) # Don't use the automatic library linking by boost with VS (#pragma ...). Instead, we specify everything here in cmake.
+  add_definitions(-DBOOST_ALL_DYN_LINK) # Link against the dynamic boost lib - needs to match with the version that find_package finds.
+endif()
+find_package(Boost 1.50.0 COMPONENTS system filesystem REQUIRED) # Why do we need boost for MorphableModel.hpp?
```
Comment (ruslo) [inline matlab/CMakeLists.txt]: We need `CONFIG` in `find_package` + `Boost::*` targets

**Comment by headupinclouds** [comment]
This one is almost ready.  This was tested with a local `HUNTER_ROOT` using a merge of the following PR's:
https://github.com/ruslo/hunter/pull/822
https://github.com/ruslo/hunter/pull/823
https://github.com/ruslo/hunter/pull/824

This PR will wait for a hunter update with the above changes.

File: matlab/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+if("${OpenCV_VERSION_MAJOR}$" EQUAL 2)
+  message(STATUS "OpenCV 2.x detected")
+  find_package(OpenCV 2.4.3 REQUIRED core imgproc highgui)
+elseif("${OpenCV_VERSION_MAJOR}$" EQUAL 3)
+  message(STATUS "OpenCV 3.x detected - including imgcodecs for compatibility")
+  find_package(OpenCV 3 REQUIRED core imgproc imgcodecs)
+endif()
+set_target_properties(${OpenCV_LIBS} PROPERTIES MAP_IMPORTED_CONFIG_RELWITHDEBINFO RELEASE)
+
+if(MSVC)
```
Comment (ruslo) [inline matlab/CMakeLists.txt]: We can remove this if we're using Hunter

File: python/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+  message(STATUS "OpenCV 2.x detected")
+  find_package(OpenCV 2.4.3 REQUIRED core imgproc)
+elseif("${OpenCV_VERSION_MAJOR}$" EQUAL 3)
+  message(STATUS "OpenCV 3.x detected - including imgcodecs for compatibility")
+  find_package(OpenCV 3 REQUIRED core imgproc imgcodecs)
+endif()
+# This allows us to compile in RelWithDebInfo. It'll use the Release-version of OpenCV:
+set_target_properties(${OpenCV_LIBS} PROPERTIES MAP_IMPORTED_CONFIG_RELWITHDEBINFO RELEASE)
+
+if(MSVC)
```
Comment (ruslo) [inline python/CMakeLists.txt]: Can be removed, see above

File: python/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+  # The standard find_package for boost on Win finds the dynamic libs, so for dynamic linking to boost we need to #define:
+  add_definitions(-DBOOST_ALL_NO_LIB) # Don't use the automatic library linking by boost with VS (#pragma ...). Instead, we specify everything here in cmake.
+  add_definitions(-DBOOST_ALL_DYN_LINK) # Link against the dynamic boost lib - needs to match with the version that find_package finds.
+endif()
+find_package(Boost 1.50.0 COMPONENTS system filesystem program_options REQUIRED)
```
Comment (ruslo) [inline python/CMakeLists.txt]: `CONFIG` see above

File: utils/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+  message(STATUS "OpenCV 2.x detected")
+  find_package(OpenCV 2.4.3 REQUIRED core imgproc highgui)
+elseif("${OpenCV_VERSION_MAJOR}$" EQUAL 3)
+  message(STATUS "OpenCV 3.x detected - including imgcodecs for compatibility")
+  find_package(OpenCV 3 REQUIRED core imgproc imgcodecs)
+endif()
+# This allows us to compile in RelWithDebInfo. It'll use the Release-version of OpenCV:
+set_target_properties(${OpenCV_LIBS} PROPERTIES MAP_IMPORTED_CONFIG_RELWITHDEBINFO RELEASE)
+
+if(MSVC)
```
Comment (ruslo) [inline utils/CMakeLists.txt]: Same

File: utils/CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+  # The standard find_package for boost on Win finds the dynamic libs, so for dynamic linking to boost we need to #define:
+  add_definitions(-DBOOST_ALL_NO_LIB) # Don't use the automatic library linking by boost with VS (#pragma ...). Instead, we specify everything here in cmake.
+  add_definitions(-DBOOST_ALL_DYN_LINK) # Link against the dynamic boost lib - needs to match with the version that find_package finds.
+endif()
+find_package(Boost 1.50.0 COMPONENTS system filesystem program_options REQUIRED)
```
Comment (ruslo) [inline utils/CMakeLists.txt]: Same

**Review by ruslo — CHANGES_REQUESTED** [review]

File: matlab/CMakeLists.txt (hunk 9 above)
Comment (headupinclouds) [reply matlab/CMakeLists.txt]: Modified (copied changes from examples folder).  Caveat: requires matlab for testing.

File: matlab/CMakeLists.txt (hunk 8 above)
Comment (headupinclouds) [reply matlab/CMakeLists.txt]: Modified (copied changes from examples folder).  Caveat: requires matlab for testing.

File: matlab/CMakeLists.txt (hunk 9 above)
Comment (headupinclouds) [reply matlab/CMakeLists.txt]: In practice, I think hunter OpenCV builds will be > 3.0.0

File: matlab/CMakeLists.txt (hunk 8 above)
Comment (headupinclouds) [reply matlab/CMakeLists.txt]: In practice, I think hunter OpenCV builds will be > 3.0.0

Comment (headupinclouds) [reply matlab/CMakeLists.txt]: Unsure about boost Windows definitions. 

```
+  add_definitions(-DBOOST_ALL_NO_LIB) # Don't use the automatic library linking by boost with VS (#pragma ...). Instead, we specify everything here in cmake.
+  add_definitions(-DBOOST_ALL_DYN_LINK) # Link against the dynamic boost lib - needs to match with the version that find_package finds.
```

File: python/CMakeLists.txt (hunk 10 above)
Comment (headupinclouds) [reply python/CMakeLists.txt]: Done.

File: python/CMakeLists.txt (hunk 11 above)
Comment (headupinclouds) [reply python/CMakeLists.txt]: Modified (copied changes from examples folder). Caveat: requires matlab for testing.

File: utils/CMakeLists.txt (hunk 13 above)
Comment (headupinclouds) [reply utils/CMakeLists.txt]: Modified (copied changes from examples folder).

File: matlab/CMakeLists.txt (hunk 8 above)
Comment (headupinclouds) [reply matlab/CMakeLists.txt]: I think we don't need this, with the cmake/hunter scheme.

**Comment by headupinclouds** [comment]
use consistent hunter_add_package for OpenCV and Boost in subdirectories
* for utils, python, matlab, examples
* assume OpenCV >= 3.0.0 w/ Hunter

```
hunter_add_package(OpenCV)
find_package(OpenCV REQUIRED core imgproc imgcodecs)

hunter_add_package(Boost COMPONENTS system filesystem program_options)
find_package(Boost CONFIG REQUIRED system filesystem program_options)
set(boost_libs Boost::system Boost::filesystem Boost::program_options)
```

**Comment by headupinclouds** [comment]
```
#!/bin/bash
CONFIG=Release
ARGS=(
    "--install "
    "--toolchain xcode"
    "--config ${CONFIG}"
    "--verbose "
    "--fwd "
    "HUNTER_CONFIGURATION_TYPES=${CONFIG}"
    "EOS_BUILD_EXAMPLES=ON"
    "EOS_BUILD_CERES_EXAMPLE=ON"    
    "EOS_BUILD_UTILS=ON"
    "EOS_GENERATE_PYTHON_BINDINGS=ON"
    "EOS_GENERATE_MATLAB_BINDINGS=ON"
)
polly.py ${ARGS[@]} --reconfig
```

`** BUILD SUCCEEDED **`

Build test looks good.

**Comment by headupinclouds** [comment]
`tree _install/xcode/`
```
_install/xcode/
├── 3rdparty
│   ├── mexplus
│   │   ├── LICENSE
│   │   └── include
│   │       ├── mexplus
│   │       │   ├── arguments.h
│   │       │   ├── dispatch.h
│   │       │   ├── mxarray.h
│   │       │   └── mxtypes.h
│   │       └── mexplus.h
│   └── pybind11
│       ├── LICENSE
│       └── include
│           └── pybind11
│               ├── attr.h
│               ├── cast.h
│               ├── chrono.h
│               ├── class_support.h
│               ├── common.h
│               ├── complex.h
│               ├── descr.h
│               ├── eigen.h
│               ├── eval.h
│               ├── functional.h
│               ├── numpy.h
│               ├── operators.h
│               ├── options.h
│               ├── pybind11.h
│               ├── pytypes.h
│               ├── stl.h
│               ├── stl_bind.h
│               └── typeid.h
├── bin  <UPDATE: removed some stuff>
│   ├── bfm-binary-to-cereal
│   ├── data
│   │   ├── image_0010.png
│   │   ├── image_0010.pts
│   │   └── notes.txt
│   ├── edgestruct-csv-to-json
│   ├── fit-model
│   ├── fit-model-ceres
│   ├── fit-model-simple
│   ├── generate-obj
│   ├── out.isomap.png
│   ├── out.mtl
│   ├── out.obj
│   ├── out.png
│   └── scm-to-cereal
├── include
│   └── eos
│       ├── core
│       │   ├── Landmark.hpp
│       │   ├── LandmarkMapper.hpp
│       │   └── Mesh.hpp
│       ├── fitting
│       │   ├── RenderingParameters.hpp
│       │   ├── affine_camera_estimation.hpp
│       │   ├── blendshape_fitting.hpp
│       │   ├── ceres_nonlinear.hpp
│       │   ├── closest_edge_fitting.hpp
│       │   ├── contour_correspondence.hpp
│       │   ├── detail
│       │   │   ├── glm_cerealisation.hpp
│       │   │   ├── nonlinear_camera_estimation_detail.hpp
│       │   │   └── optional_cerealisation.hpp
│       │   ├── fitting.hpp
│       │   ├── linear_shape_fitting.hpp
│       │   ├── nonlinear_camera_estimation.hpp
│       │   └── orthographic_camera_estimation_linear.hpp
│       ├── morphablemodel
│       │   ├── Blendshape.hpp
│       │   ├── EdgeTopology.hpp
│       │   ├── MorphableModel.hpp
│       │   ├── PcaModel.hpp
│       │   ├── coefficients.hpp
│       │   └── io
│       │       ├── cvssp.hpp
│       │       ├── eigen_cerealisation.hpp
│       │       └── mat_cerealisation.hpp
│       ├── render
│       │   ├── FragmentShader.hpp
│       │   ├── Rasterizer.hpp
│       │   ├── SoftwareRenderer.hpp
│       │   ├── VertexShader.hpp
│       │   ├── detail
│       │   │   ├── Vertex.hpp
│       │   │   ├── render_affine_detail.hpp
│       │   │   ├── render_detail.hpp
│       │   │   └── texture_extraction_detail.hpp
│       │   ├── render.hpp
│       │   ├── render_affine.hpp
│       │   ├── texture_extraction.hpp
│       │   └── utils.hpp
│       └── video
│           └── Keyframe.hpp
├── matlab
│   ├── +eos
│   │   ├── +fitting
│   │   │   ├── fit_shape_and_pose.m
│   │   │   └── private
│   │   │       └── fitting.mexmaci64
│   │   └── +render
│   │       ├── extract_texture.m
│   │       ├── private
│   │       │   └── render.mexmaci64
│   │       └── render.m
│   ├── demo.m
│   └── include
│       ├── mexplus_eigen.hpp
│       ├── mexplus_eos_types.hpp
│       └── mexplus_opencv.hpp
├── python
│   ├── demo.py
│   └── eos.so
└── share
    ├── convert_bfm2009_to_raw_binary.m
    ├── expression_blendshapes_3448.bin
    ├── ibug_to_sfm.txt
    ├── model_contours.json
    ├── readme.txt
    ├── reference.obj
    ├── reference_annotated.obj
    ├── reference_symmetry.txt
    ├── sfm_3448_edge_topology.json
    └── sfm_shape_3448.bin
```

**Comment by ruslo** [comment]
```
├── bin
│   ├── _builds
│   │   └── xcode
│   │       └── _3rdParty
│   │           └── polly
│   ├── _logs
│   │   └── polly
│   │       └── xcode
│   │           └── log.txt
```

in install?

File: matlab/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+  # DOCUMENTATION +eos/+fitting/fit_shape_and_pose.m # doesn't work - wrong path probably. But it renames the file to fitting.m, so not what we want anyway.
+  LINK_TO eos ${OpenCV_LIBS} ${boost_libs} #[LINK_TO target1 target2 ...]
+  #[...]
+)
+
+matlab_add_mex(
+  NAME eos-matlab-render
+  SRC +eos/+render/private/render.cpp
+  OUTPUT_NAME render
+  LINK_TO eos ${OpenCV_LIBS} ${Boost_LIBRARIES}
```
Comment (ruslo) [inline matlab/CMakeLists.txt]: Boost_LIBRARIES -> boost_libs

**Review by ruslo — CHANGES_REQUESTED** [review]

File: matlab/CMakeLists.txt (hunk 14 above)
Comment (headupinclouds) [reply matlab/CMakeLists.txt]: Good catch.

**Comment by headupinclouds** [comment]
> in install?

I just ran it clean build + install.  Looks fine.

**Review by ruslo — APPROVED** [review]
CMake part looks good for me

**Comment by headupinclouds** [comment]
👍 Thanks for the review + fixes and quick turnaround w/ hunter dependencies.  I'm testing locally w/ the updated hunter now.  I guess per package .travis.yml and appveyor.yml tests can be added.  I believe the current recommendation is to add per package CI tests here `ingenue/hunter/pkg.eos`?

**Comment by headupinclouds** [comment]
For reference: https://github.com/ruslo/hunter/issues/743

**Comment by ruslo** [comment]
> I believe the current recommendation is to add per package CI tests here ingenue/hunter/pkg.eos?

Let's test that this version doesn't broke current config, if so later after merge we can extend `.travis.yml`/`appveyor.yml` of `eos` package.

**Comment by headupinclouds** [comment]
Ok.  The local build via HunterGate worked fine.  I've tagged a release and will create a hunter PR.

**Comment by headupinclouds** [comment]
https://github.com/ruslo/hunter/pull/825 🚀

**Pull request merged**